src/Cornelis/Offsets.hs

```diff
@@ -1,27 +1,139 @@
 import qualified Data.Text as T
 import           Data.Text.Encoding (encodeUtf8)
+import           GHC.Generics (Generic)
+import           GHC.Stack (HasCallStack)
+import           Prettyprinter (Pretty)
 
+-- TODO: Distinguish offsets and positions; offsets have arithmetic, positions don't
+
+-- | Indexing scheme: whether the first index is zero or one.
+data Indexing = OneIndexed | ZeroIndexed
+
+-- | What are we counting?
+data Unit = Byte | CodePoint | Line
+
+-- | The constructor is hidden, use 'zeroIndexed' and 'oneIndexed' to construct it,
+-- and 'fromZeroIndexed' and 'fromOneIndexed' to destruct it.
+newtype Index (e :: Unit) (i :: Indexing) = Index Int
+  deriving newtype (Eq, Ord, Show, Read, FromJSON, Pretty)
+
+-- | It doesn't seem worth the trouble to hide this constructor.
+newtype Offset (e :: Unit) = Offset Int
+  deriving newtype (Eq, Ord, Show, Read, FromJSON, Pretty)
+
+-- | Position in a text file as line-column numbers. This type is indexed by
+-- the units of the columns (@Byte@ or @CodePoint@) and by the indexing scheme
+-- of lines and columns.
+data Pos e i j = Pos
+  { p_line :: Index 'Line i
+  , p_col :: Index e j
+  } deriving (Eq, Ord, Show, Generic)
+
+data Interval p = Interval { iStart, iEnd :: !p }
+  deriving (Eq, Ord, Show, Functor, Foldable, Traversable, Generic)
+
+-- Common specializations
+
+type LineNumber = Index 'Line
+
+type AgdaIndex = Index 'CodePoint 'OneIndexed
+type AgdaOffset = Offset 'CodePoint
+type AgdaPos = Pos 'CodePoint 'OneIndexed 'OneIndexed
+type AgdaInterval = Interval AgdaPos
+
+type VimIndex = Index 'Byte
+type VimOffset = Offset 'Byte
+type VimPos = Pos 'Byte 'ZeroIndexed 'ZeroIndexed
+type VimInterval = Interval VimPos
+
+-- To pass indices to and from external sources.
+
+-- | Mark a raw index as zero-indexed.
+zeroIndexed :: Integral a => a -> Index e 'ZeroIndexed
+zeroIndexed a = Index (fromIntegral a)
+
+-- | Mark a raw index as one-indexed.
+oneIndexed :: Integral a => a -> Index e 'OneIndexed
+oneIndexed a = Index (fromIntegral a)
+
+-- | Unwrap a raw zero-indexed index.
+fromZeroIndexed :: Num a => Index e 'ZeroIndexed -> a
+fromZeroIndexed (Index a) = fromIntegral a
+
+-- | Unwrap a raw zero-indexed index.
+fromOneIndexed :: Num a => Index e 'OneIndexed -> a
+fromOneIndexed (Index a) = fromIntegral a
+
+-- | Convert from one- to zero-indexed.
+zeroIndex :: Index e 'OneIndexed -> Index e 'ZeroIndexed
+zeroIndex (Index i) = Index (i - 1)
+
+-- | Convert from zero- to one-indexed.
+oneIndex :: Index e 'ZeroIndexed -> Index e 'OneIndexed
+oneIndex (Index i) = Index (i + 1)
+
+-- | Increment index.
+incIndex :: Index e i -> Index e i
+incIndex (Index i) = Index (i + 1)
+
+-- | Add an offset to an index.
+(.+) :: Index e i -> Offset e -> Index e i
+Index i .+ Offset n = Index (i + n)
+
+(.-.) :: Index e i -> Index e i -> Offset e
+Index i .-. Index j = Offset (i - j)
+
+--
 
 offsetPlus :: Offset a -> Offset a -> Offset a
-offsetPlus = coerce $ (+) @Int32
```

Comment:
I hope it's okay to remove this `Int32`

Reply:
no problem; I think this originally started as something specialized for vim